### DISCUSSION TITLE
time zone problem about nanosToWaitForRefill and nanosToWaitForReset  fields.

### DISCUSSION THREAD

**Discussion by ss-guvenc** [vladimir-bukhtoyarov/bucket4j#211] [Q&A] [answered]
Hi,
I am trying to implement rate limiter with bucked4j on spring boot.

I am using; 
spring boot 2.6.1
hazelcast 5.0.2
bucket4j 7.0.0

i can send my backet state to hazelcast. But there is problem about time info in ConsumptionProbe class.
Because of this, my created bucket is never refill. And my rest call is blocked by rate limiter.

i think  this problem about nanosToWaitForRefill and nanosToWaitForReset fields in ConsumptionProbe class.
The time on these fields is not properly set.

Just an example ;
client time: Mon Dec 27 08:33:00 UTC 2021
nanosToWaitForRefill: Fri Nov 26 10:39:36 UTC 1971
nanosToWaitForReset: Fri Nov 26 10:39:36 UTC 1971

Thanks for your help

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @ss-guvenc 

nanosWaitToRefill does not point to time instant, instead it describes the interval of time that needs to wait before requested amount of tokens can be consumed.

**Reply by ss-guvenc** [reply]
my rate limiter test code illustrated below.
```
    public Boolean tryConsumeToken(String apiKey) {
        RemoteCommand<ConsumptionProbe> tryConsumeCommand = new TryConsumeAndReturnRemainingTokensCommand(1);
        CommandResult commandResult = proxyManager.execute(apiKey, new Request(tryConsumeCommand, Versions.v_7_0_0, System.currentTimeMillis()));
        if (commandResult.isBucketNotFound()) {
            return createInitialBucket(apiKey);
        } else {
            ConsumptionProbe consumptionProbe = (ConsumptionProbe) commandResult.getData();
            if (consumptionProbe.isConsumed()) {
                return true;
            } else {
                return false; // after consuming all tokens. always return false here.
            }
        }
    }

    public Boolean createInitialBucket(String apiKey) {
        CreateInitialStateCommand createInitialStateCommand = new CreateInitialStateCommand(newBucketConfiguration());
        proxyManager.execute(apiKey, new Request(createInitialStateCommand, Versions.v_7_0_0, System.currentTimeMillis()));

        RemoteCommand<Boolean> tryConsumeCommand = new TryConsumeCommand(1);
        CommandResult createInitialStateAndExecuteCommandResult = proxyManager.execute(apiKey, new Request(tryConsumeCommand, Versions.v_7_0_0, System.currentTimeMillis()));
        return (Boolean) createInitialStateAndExecuteCommandResult.getData();
    }

    private BucketConfiguration newBucketConfiguration() {
        return BucketConfiguration.builder()
                .addLimit(Bandwidth.classic(capacity, Refill.intervally(capacity, Duration.ofSeconds(refillInterval))))
                .build();
    }
```
 First token is created but never refill. 

 After consuming all tokens. always return false.

 Am i doing something wrong?

capacity = 3 , refillInterval=60

thx @vladimir-bukhtoyarov

**Reply by vladimir-bukhtoyarov** [reply]
You are trying to work trough low-level API. ProxyManager#execute is not supposed to be used by clients. Please stay on Bucket abstraction. ProxyManager has builder api for buckets, example of usage can be found there https://bucket4j.com/7.0.0/toc.html#example-1-limiting-access-to-http-server-by-ip-address

**Reply by ss-guvenc** [reply]
thx @vladimir-bukhtoyarov
